client/src/legacy/LeftAndMain.EditForm.js

```diff
@@ -566,22 +566,74 @@ $.entwine('ss', function($){
     onmatch: function () {
       if (this.val() === 'OnlyTheseUsers') {
         if (this.is(':checked')) {
-          this.showList(true);
+          this.showGroupsList(true);
+          this.hideMembersList(true);
         } else {
-          this.hideList(true);
+          this.hideGroupsList(true);
+        }
+      }
+
+      if (this.val() === 'OnlyTheseMembers') {
+        if (this.is(':checked')) {
+          this.showMembersList(true);
+          this.hideGroupsList(true);
+        } else {
+          this.hideMembersList(true);
         }
       }
     },
     onchange: function (e) {
       if (e.target.value === 'OnlyTheseUsers') {
-        this.showList();
+        this.showGroupsList();
+        this.hideMembersList();
+      } else {
+        this.hideGroupsList();
+      }
+      if (e.target.value === 'OnlyTheseMembers') {
+        this.showMembersList();
+        this.hideGroupsList();
+      } else {
+        this.hideMembersList();
+      }
+    },
+    // these two functions retained for backwards compatibility
+    showList: function(instant) {
+      this.showGroupsList(instant);
+      this.showMembersList(instant);
+    },
+    hideList: function(instant) {
+      this.hideGroupsList(instant);
+      this.hideMembersList(instant);
+    },
+    showGroupsList: function (instant) {
+      const holder = this.closest('.field');
+      const list = holder.next().filter('.treemultiselect');
+      holder.addClass('field--merge-below');
+      if (instant) {
+        list.show().css('overflow', 'visible');
       } else {
-        this.hideList();
+        list.slideDown(() => {
+          // jquery adding overflow:hidden in hide, this will break listbox display
+          list.css('overflow', 'visible');
+        });
+      }
+    },
+    hideGroupsList: function (instant) {
+      const holder = this.closest('.field');
+      const list = holder.next().filter('.treemultiselect');
+      list.css('overflow', 'hidden');
+      if (instant) {
+        list.hide().css('display', 'none');
+        holder.removeClass('field--merge-below');
+      } else {
+        list.slideUp(() => {
+          holder.removeClass('field--merge-below');
+        });
       }
```

Comment:
Given that this is all duplicated from `hideMembersList()`, can you please migrate that shared functionality into a shared function, so we can have DRY code?
Same with the `showGroupsList()` and `showMembersList()` shared functionality.